lib/safe-url.js

```diff
@@ -0,0 +1,75 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// This module exports a safe URL-builder interface, ensuring that no
+// unsafe input can leak into generated URLs.
+//
+// It takes the approach of throwing error.unexpectedError() when unsafe
+// input is encountered, for extra visibility. An alternative approach
+// would be to use encodeURIComponent instead to convert unsafe input on
+// the fly. However, we have no valid use case for encoding weird data
+// like that, since we explicitly hex-encode params that need it. So if
+// any weird input is encountered, we should fail the request as soon as
+// possible.
+//
+// Usage:
+//
+//   const SafeUrl = require('./safe-url')(log)
+//
+//   const url = new SafeUrl('/account/:uid/sessions', 'db.sessions')
+//   url.render({ uid: 'foo' })            // returns '/account/foo/sessions'
+//   url.render({ uid: 'bar' })            // returns '/account/bar/sessions'
+//   url.render({ uid: 'foo\n' })          // throws error.unexpectedError()
+//   url.render({})                        // throws error.unexpectedError()
```

Comment:
nit: Should we throw on `{}` if no path params defined?

Reply:
I don't think so, because it makes calling easier for those cases. And that's what the current behaviour is too. But I'm not sure if you're saying it's wrong that we don't throw in that case, or you just didn't realise that we already aren't?

Reply:
@philbooth This does perform as you stated. It will not throw with `url.render({})` if a path with no params are specified, ex. `/foo/bar`.